mkfat32.zig

```diff
@@ -0,0 +1,361 @@
+    ///     16GB   -> 32GB:  16384 bytes (16KB)
+    ///     32GB   -> 2TB:   32768 bytes (32KB)
+    ///     else:           Error
+    ///
+    /// The smallest size is calculated as follows:
+    ///     (reserved_sectors + num_of_fats + root_directory) * smallest_sector_size = (32 + 2 + 1) * 512 = 35 * 512
+    ///
+    /// Arguments:
+    ///     IN image_size: u64       - The image size the user wants to create.
+    ///     IN bytes_per_sector: u16 - The bytes per sector the user defined, or the default value.
+    ///
+    /// Return: u16
+    ///     The default sectors per cluster based on the image size and sector size.
+    ///
+    /// Error: Error
+    ///     error.TooSmall - If the image size is too small (below 35 * 512 (17.5KB))
+    ///     error.TooLarge - If the image size is too large (above 2TB)
+    ///
+    pub fn getDefaultSectorsPerCluster(image_size: u64, bytes_per_sector: u16) Error!u8 {
+        // A couple of constants to help
+        const KB = 1024;
+        const MB = 1024 * KB;
+        const GB = 1024 * MB;
+        const TB = 1024 * GB;
+
+        return @intCast(u8, try switch (image_size) {
+            0...35 * 512 - 1 => Error.TooSmall,
+            35 * 512...64 * MB - 1 => std.math.max(512, bytes_per_sector) / bytes_per_sector,
+            64 * MB...128 * MB - 1 => std.math.max(1024, bytes_per_sector) / bytes_per_sector,
+            128 * MB...256 * MB - 1 => std.math.max(2048, bytes_per_sector) / bytes_per_sector,
+            256 * MB...8 * GB - 1 => std.math.max(4096, bytes_per_sector) / bytes_per_sector,
+            8 * GB...16 * GB - 1 => std.math.max(8192, bytes_per_sector) / bytes_per_sector,
+            16 * GB...32 * GB - 1 => std.math.max(16384, bytes_per_sector) / bytes_per_sector,
+            32 * GB...2 * TB - 1 => std.math.max(32768, bytes_per_sector) / bytes_per_sector,
+            else => Error.TooLarge,
+        });
+    }
+
+    ///
+    /// Get the default volume name of a FAT image: "NO NAME    ".
+    ///
+    /// Return: [11]u8
+    ///     The default volume name.
+    ///
+    pub fn getDefaultVolumeName() [11]u8 {
+        return "NO NAME    ".*;
+    }
+
+    ///
+    /// Get the default image size: 33548800 (~32KB). This is the recommended minimum image size
+    /// for FAT32. 0xFFF5 (valid cluster values) * 512 (getDefaultSectorSize).
+    ///
+    /// Return: u32
+    ///     The smallest recommended FAT32 image size.
+    ///
+    pub fn getDefaultImageSize() u32 {
+        return 0xFFF5 * @as(u32, getDefaultSectorSize());
+    }
+
+    ///
+    /// Create the FAT32 serial number. This is generated from the date. Reference:
+    /// https://www.digital-detective.net/documents/Volume%20Serial%20Numbers.pdf
+    ///
+    /// Arguments:
+    ///     IN current_time: SerialTime - The current date and time.
+    ///
+    /// Return: u32
+    ///     Serial number for a new FAT32 image.
+    ///
+    pub fn createSerialNumber(current_time: SerialTime) u32 {
```

Comment:
This could be moved to the SerialTime struct for nice encapsulation.

Reply:
I don't think it should be moved as it is a static function and there isn't a need for `Self` for `SerialTime`. It would encapsulate it but would make the use of it longer: `Fat32.createSerialNumber()` =>`Fat32.SerialTime.createSerialNumber()`. Also will need to make the struct public.

Reply:
It's static for now, but if you have a `SerialTime` instance, you'd just do `serial_time.createSerialNumber()` instead of `Fat32.createSerialNumber(serial_time)`, so it would actually be shorter!

Reply:
There wouldn't be any other need for SerialTime other than creating the serial number. The only reason I have the SerialTime struct was to reduce the number of parameters being passed to the `createSerialNumber`. It would just create code blot and be more complex than needed. Once there is std lib get human readable time, I can have another look at changing how this is done. Probable removing the the argument for `createSerialNumber`, removing the `SerialTime` struct, and having the function call the std lib function directly.

Reply:
Hmm, I still think that putting the function in the struct is better than having it in Fat32, since that's how we and the std lib do things in other structs. I don't agree with the code bloat argument since it won't mean adding more lines and it would actually be cleaner!. Alternatively I don't mind having 6 parameters for the function.

Reply:
I shall do parameters